### ISSUE TITLE
Invalid type Inference when generic function passed as argument in another generic

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#6027]
I found problem when generic callable passed into another generic function without intermediate variable:

```python
T = TypeVar("T")

def generic_proxy() -> Callable[[T], T]:
    return lambda x: x

items: list[int] = [1]

for item in map(generic_proxy(), items):
    reveal_type(item) # Reveal as "item: _T1@__init__", but must be int.
```

Variable for callable will fix problem:

```python
proxy = generic_proxy()
for item in map(proxy, items):
    reveal_type(item) # Reveal as int.
```

pyright 1.1.327

**Comment by erictraut** [comment]
Thanks for the bug report. It turns out that this issue shares the same root cause as #6042, which I just fixed. So this issue will also be fixed in the next release.

Interestingly, mypy also exhibits the same bug in this case.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.331, which I just published. It will also be included in a future release of pylance.
